### DISCUSSION TITLE
CamBaseCamera class

### DISCUSSION THREAD

**Discussion by kagikn** [scripthookvdotnet/scripthookvdotnet#1252] [Ideas]
I'm considering adding the `CamBaseCamera` or `BaseCamera` class so you can fetch camera info from the active camera in the game, but NOT limited to scripted camera like how `GET_RENDERING_CAM` returns. Any questions, suggestions, or potentially ideas? Please refrain from suggesting what new features SHVDN camera classes should have without stuff obtained by reverse engineering the exe in this discussion.

There's `camBaseCamera` and the pool for `camBaseCamera` instance. The game uses pool handles for scripted camera handles, just like how `CEntity` handles are created and how we use entity (strictly `CPhysical`) handles via native functions, but the game also stores gameplay camera instance (not `camBaseDirector` ones) to the same pool. 
The problem is, not all existing members of `Camera` class can use with non-scripted camera handles. Some of the camera natives that takes a cam handle can work properly only with scripted cam handles since they also access the `camScriptDirector` instance or some other script cam variables, such as `SET_CAM_ACTIVE`. While others such as `GET_CAM_FAR_DOF` can work with arbitrary cam handles without any potential issues such as pointer malformation, meaning they access only the members of `camBaseCamera`. That is why I prefer the class for `camBaseCamera` as a new class over treating existing `Camera` class as the base cam class in v3 API, although you won't be able to recognize existing `Camera` class as scripted cams too quickly. We'll make a class like `ScriptedCamera` or `CamScriptedCamera` so you can recognize the scripted camera class in v4, though.

We can introduce the new `CamBaseCamera` as the base class of existing `Camera`, but I believe we shouldn't add `sealed` keyword for the new class so we can add subclasses that represents some of the subclasses of `camBaseCamera` such as `camFollowPedCamera`.

For those wondering how to fetch the `camBaseCamera` pool address and the functions for active `camBaseCamera` address:
```cs
IntPtr patternAddr = Game.FindPattern("45 8B C8 4D 85 DB 7E 59 49 8B 42 08 42 0F B6 14 08");
IntPtr ptrToptrToCamBaseCameraPool = (patternAddr + *(int*)(patternAddr - 11) - 7); // use GenericPool interface in NativeMemory to mess with
patternAddr = Game.FindPattern("44 0F 29 40 A8 E8 ? ? ? ? 45 33 ED 48 8B D8 48 85 C0 0F 84");
(delegate* unmanaged[Stdcall]<IntPtr>)getActiveCamera = (delegate* unmanaged[Stdcall]<IntPtr>)(patternAddr + *(int*)(patternAddr + 6) + 10); // gets the active camBaseCamera address
```

**Comment by choiszt** [comment]
Hi @kagikn !
I'm currently attempting to align point cloud data from RGBD collected in GTA-V. However, I'm struggling to find the appropriate intrinsic matrices. Moreover, I've been unable to locate the extrinsic matrix necessary to accurately transform the depth image to world coordinates. I suspect the depth images might not be in the camera coordinates, which means I can't directly apply the camera parameters for the transformation. Are there any chances in your latest powerful scripthook to acquire the correct **intrinsic matrix** of default camera and the **transformation matrix of from depth coords to World**?(I guess I miss the NDC to World matrix)  Any insights or experience you could share would be greatly appreciated. Thank you in advance

**Reply by kagikn** [reply]
Hi!
I am afraid that your question does not fit into this discussion. This discussion is about camera classes with stuff obtained by reverse engineering the exe. Since you have created a dedicated issue in #1335 (though I converted into a discussion), we can talk in there.

**Comment by kagikn** [comment]
We know what `camBaseCamera` has now. Maybe more stuff have been added since a build in 2022 though. And no, `camBaseCamera` doesn't have any info about lookat or attach parent info except for a lookat target entity or an attach parent (`camScriptedCamera` has).

This is what `camBaseCamera` has as its data:
```cpp
class camBaseCamera : public camBaseObject {
    int m_NumUpdatesPerformed; // 0x20
    camFrame *m_Frame; // 0x30
    camFrame *m_PostEffectFrame; // 0x110
    rage::atArray<rage::fwRegdRef<camBaseFrameShaker,rage::fwRefAwareBase>,0,unsigned short> m_FrameShakers; // 0x1F0
    uintptr_t m_FrameInterpolator; // 0x200
    uintptr_t m_DofSettings; // 0x208
    rage::fwRegdRef<camCollision,rage::fwRefAwareBase> m_Collision; // 0x210
    uintptr_t m_Metadata; // 0x218
    rage::fwRegdRef<CEntity const ,rage::fwRefAwareBase> m_AttachParent; // 0x220
    rage::fwRegdRef<CEntity const ,rage::fwRefAwareBase> m_LookAtTarget; // 0x228
    // _bf230 at 0x230
}; // size: 0x240
```

`camBaseObject` definition for its data (we don't have no real useful members in here):
```cpp
class camFrame : public rage::fwRefAwareBase {
    uintptr_t m_Metadata; // 0x10
    uintptr_t m_WidgetGroup; // 0x18
}; // size: 0x20
```

For those wondering what `fwRegdRef` does, it holds a reference. `rage::fwRefAwareBaseImpl<rage::datBase>`, which is the direct direct base class of `rage::fwRefAwareBase`, holds known references and drop its stuff when it stops tracking any references.
 
`camFrame` definition for its data:
```cpp
class camFrame : public rage::fwRefAwareBase {
    rage::Matrix34 m_WorldMatrix; // 0x10
    rage::Vector4 m_DofPlanes; // 0x50
    rage::Vector2 m_DofFocusDistanceGridScaling; // 0x60
    rage::Vector2 m_DofSubjectMagPowerNearFar; // 0x68
    float m_Fov; // 0x70
    float m_NearClip; // 0x74
    float m_FarClip; // 0x78
    float m_CameraTimeScaleThisUpdate; // 0x7c
    float m_DofStrength; // 0x80
    float m_DofBlurDiskRadius; // 0x84
    float m_DofMaxPixelDepth; // 0x88
    float m_DofMaxPixelDepthBlendLevel; // 0x8C
    float m_DofPixelDepthPowerFactor; // 0x90
    float m_FNumberOfLens; // 0x94
    float m_FocalLengthMultiplier; // 0x98
    float m_DofFStopsAtMaxExposure; // 0x9C
    float m_DofFocusDistanceBias; // 0xA0
    float m_DofMaxNearInFocusDistance; // 0xA4
    float m_DofMaxNearInFocusDistanceBlendLevel; // 0xA8
    float m_DofMaxBlurRadiusAtNearInFocusLimit; // 0xAC
    float m_DofBlurDiskRadiusPowerFactorNear; // 0xB0
    float m_DofOverriddenFocusDistance; // 0xB4
    float m_DofOverriddenFocusDistanceBlendLevel; // 0xB8
    float m_DofFocusDistanceIncreaseSpringConstant; // 0xBC
    float m_DofFocusDistanceDecreaseSpringConstant; // 0xC0
    float m_DofPlanesBlendLevel; // 0xC4
    float m_DofFullScreenBlurBlendLevel; // 0xC8
    float m_MotionBlurStrength; // 0xCC
    rage::fwFlags<unsigned short> // 0xD0
}; // size: 0xE0
```
